### PR TITLE
extmod/modubinascii: implement binascii.crc32

### PR DESCRIPTION
**Pull request by prusnak**
as per https://docs.python.org/3/library/binascii.html#binascii.crc32

### CODE REVIEW

**Comment by prusnak** [comment]
as per https://docs.python.org/3/library/binascii.html#binascii.crc32

**Comment by pfalcon** [comment]
This is very large implementation. Try to beat in size the one in uzlib.

**Comment by dhylands** [comment]
> This is very large implementation. Try to beat in size the one in uzlib.

As far as I can tell, uzlib uses adler-32 which is a totally different algorithm.

The only mention of crc32 in uzlib is in tinf.h which has a prototype for a function which isn't present.

I think what @pfalcon meant to say, was to take a look at this implementation here:
https://bitbucket.org/jibsen/tinf/src/d4327ed5fe3826620e2c53c292d456d5cb6b5932/src/crc32.c?at=default&fileviewer=file-view-default

which only uses a 16-entry lookup table (so it trades off performance for amount of flash/RAM used).

I'd also recommend that you expose the crc function using a C API so that we can use it from elsewhere.

**Comment by dhylands** [comment]
It would also be good to include some tests for the crc function to verify that its behaving properly.

**Comment by prusnak** [comment]
I can provide two implementations - one optimized for speed (this one)
and one optimized for size. And a compile time option to choose
between the too.

**Comment by prusnak** [comment]
Wrt the tests: I would appreciate some pointers what is the best way
to do it. Do you have some testing framework already in place? I
don't see tests for other ubinascii functions where I could get
inspiration from ...

**Comment by dhylands** [comment]
There are test files in the tests directory tree. Here's an example for binascii.unhexlify:
https://github.com/micropython/micropython/blob/master/tests/extmod/ubinascii_unhexlify.py

You can run the test suite by building the unix version of micropython and then from the tests directory issuing the ```./run-tests``` or run a single test by doing: ```./run-tests extmod/ubinascii_unhexlify.py```

The way that the tests work is that they'll run the file using cpython and then run the test using micropython. This will create a .exp file (expected - from CPyton) and an .out file (from MicroPython). If the files are the same, then the output files are removed. If the files are different, then the .exp and .out files will be left behind (so you can see what's not working).

It's also possible to run the test on a pyboard, and for some cases, we can't run CPython since its testing a MicroPython only feature. In that case the .exp file will be checked in.

File: extmod/modubinascii.c [hunk 1]
```diff
@@ -203,6 +203,68 @@ mp_obj_t mod_binascii_b2a_base64(mp_obj_t data) {
 }
 MP_DEFINE_CONST_FUN_OBJ_1(mod_binascii_b2a_base64_obj, mod_binascii_b2a_base64);
 
+static uint32_t crc32_table[] = {
```
Comment (dhylands) [inline extmod/modubinascii.c]: This table should be declared const so that it goes in flash instead of RAM

**Comment by pfalcon** [comment]
> The only mention of crc32 in uzlib is in tinf.h which has a prototype for a function which isn't present.

https://github.com/pfalcon/uzlib/blob/master/src/crc32.c

**Comment by dhylands** [comment]
> The only mention of crc32 in uzlib is in tinf.h which has a prototype for a function which isn't present.
> 
> https://github.com/pfalcon/uzlib/blob/master/src/crc32.c

You didn't specify which uzlib, and that file isn't present in the micropython tree:
https://github.com/micropython/micropython/tree/master/extmod/uzlib

**Comment by prusnak** [comment]
Will update PR tomorrow.

**Comment by prusnak** [comment]
Is copypasting the function and lookup table from crc32.c to
modubinascii.c acceptable?

**Comment by dhylands** [comment]
You can view the license at the bottom of this page:
https://bitbucket.org/jibsen/tinf/src/d4327ed5fe38?at=default

so as long as you provide appropriate attribution, then you should be fine.

**Comment by prusnak** [comment]
Updated PR:
* Replaced code with the one from uzlib.
* Added const to lookup table.
* Added ubinascii_crc32 test.

**Comment by pfalcon** [comment]
Please avoid duplicating code which is already present in libs integrated into uPy (unless you have arguments for doing otherwise). In this particular case, please use crc32 function from uzlib.

**Comment by prusnak** [comment]
@pfalcon You added crc32 code to `extmod/uzlib` way after this pull request was created. I will rework the code, but I suggest you quit making fools of micropython contributors in the future.

**Comment by prusnak** [comment]
Reworked in commit 39799f7564d28737fa815a3a5d370c91414998ef

**Comment by pfalcon** [comment]
> You added crc32 code to extmod/uzlib way after this pull request was created. I will rework the code, but I suggest you quit making fools of micropython contributors in the future.

You were suggested to do that at the very beginning. You ignored it, though nothing precluded you to add it as part of your patch. And please leave such attitude behind: if you don't don't do it right, you will be asked to re-do it right as many times as it takes you to do it - in any projects which cares a little bit about code quality. Try to contribute to Linux or something, and then come back and see how welcoming and relaxing we have it. Though ignoring *suggestions* won't get you far still - you will be *asked* to re-do it.

**Comment by deshipu** [comment]
@pfalcon Please let's not take Linus as a role model here. It's a very dysfunctional community. It makes a big difference how you word the comment, and I believe that's what @prusnak is complaining about, not the necessity to update the code. Right now you are the gate keeper of this project and you can keep offending people, but do it too much, and they will go to a different project or even simply make a fork -- neither is good for this community.

**Comment by prusnak** [comment]
As you can see I have no problems with reworking the patch as I did it 3 times already.

Also about the copy&paste issue, I asked on 19th of July and it was acked by Dave a comment later.

Pointing out that some community is more hostile than this one is a fallacy argument. You can always find shit that smells more than yours, but that doesn't mean yours is fine. I can also play hardball and I am good at it, but I prefer not to.

**Comment by pfalcon** [comment]
"Linus" is not related here. It's just any big enough project at some point reaches big enough churn that it either accepts any code and becomes Augean Stables, so even original creators lose interest in it, or it remains careful about code, but with increased churn, it all becomes more complicated.

As for "offending", I don't see anything offending in asking to rework the patch. Definitely not more offending as me hinting at the way not to duplicate code and him ignoring it and proceeding to duplicate.

Generally, there's hundred of people contributed to MicroPython, and only few make a fuzz. Gentlemen, please take it easy and concentrate on important matters (like cooperation on improving MicroPython).

**Comment by prusnak** [comment]
Agree. Let's get this change in or describe why it can't go in.

**Comment by pfalcon** [comment]
> As you can see I have no problems with reworking the patch as I did it 3 times already.

Good, thanks. Rest assured that other people (maintainers included) rework their stuff not less.

> Also about the copy&paste issue, I asked on 19th of July and it was acked by Dave a comment later.

I can't speak for Dave, nor can't tell you: "Listen to me, not Dave". So, please use your best judgement when deciding what to do.

> Pointing out that some community is more hostile than this one is a fallacy argument. You can always find shit that smells more than yours, but that doesn't mean yours is fine. I can also play hardball and I am good at it, but I prefer not to.

For reference sake, https://github.com/git-multimail/git-multimail/pull/162 is an example of a patch where I "couldn't make it" (i.e. couldn't produce a correctly working patch, which would satisfy maintainer) or https://gerrit.zephyrproject.org/r/#/c/3895/ is where my patch was closed simply because it was submitted in "wrong time", asking to resubmit it again (apparently, even without changes). So, stuff like someone not taking your patches, even for funny (to you) reasons, happens all the time.

Now back to your passage: calling some community hostile is well, hostile. Even more so "shitting" around. Please don't do that. This thread is now closed. (The change gets in, thanks, unless submitter will withdraw it till the end of day).

**Comment by pfalcon** [comment]
But wait, as you can see, this change breaks a build.

**Comment by pfalcon** [comment]
(I'm on it.)

**Comment by pfalcon** [comment]
Fixed in c4283675439d549fac09c5695aa5f8d6c3319dd5. But shows there's no shortage of things to re/up-work in submitted patches. I leave hint about updating docs to @deshipu . Thanks for contribution!

**Comment by prusnak** [comment]
Thanks!

**Pull request merged**